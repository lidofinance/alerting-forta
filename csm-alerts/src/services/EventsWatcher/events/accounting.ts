import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { CSAccounting__factory } from '../../../generated/typechain'
import * as CSAccounting from '../../../generated/typechain/CSAccounting'
import { EventOfNotice } from '../../../shared/types'
import { etherscanAddress, formatEther } from '../../../utils/string'

const ICSAccounting = CSAccounting__factory.createInterface()

export function getCSAccountingEvents(accounting: string): EventOfNotice[] {
    return [
        {
            address: accounting,
            abi: ICSAccounting.getEvent('ChargePenaltyRecipientSet').format('full'),
            alertId: 'CS-ACCOUNTING-CHARGE-PENALTY-RECIPIENT-SET',
            name: 'ðŸš¨ CSAccounting: Charge penalty recipient set',
            description: (args: CSAccounting.ChargePenaltyRecipientSetEvent.OutputObject) =>
                `Charge penalty recipient set to ${etherscanAddress(args.chargePenaltyRecipient)} (expecting the treasury)`,
            severity: FindingSeverity.Critical,
            type: FindingType.Info,
        },
        {
            address: accounting,
            abi: ICSAccounting.getEvent('BondCurveUpdated').format('full'),
            alertId: 'CS-ACCOUNTING-BOND-CURVE-UPDATED',
            name: 'ðŸš¨ CSAccounting: Bond curve updated',
            description: (args: CSAccounting.BondCurveUpdatedEvent.OutputObject) => {
                const bondCurveString = args.bondCurve
                    .map((value: bigint) => formatEther(value))
                    .join(', ')
                return `Bond curve #${args.curveId} updated. New values: [${bondCurveString}]`
            },
            severity: FindingSeverity.Critical,
            type: FindingType.Info,
        },
        {
            address: accounting,
            abi: ICSAccounting.getEvent('BondCurveAdded').format('full'),
            alertId: 'CS-ACCOUNTING-BOND-CURVE-ADDED',
            name: 'ðŸ”´ CSAccounting: Bond curve added',
            description: (args: CSAccounting.BondCurveAddedEvent.OutputObject) => {
                const bondCurveString = args.bondCurve
                    .map((value: bigint) => formatEther(value))
                    .join(', ')
                return `Bond curve added: [${bondCurveString}]`
            },
            severity: FindingSeverity.High,
            type: FindingType.Info,
        },
    ]
}
