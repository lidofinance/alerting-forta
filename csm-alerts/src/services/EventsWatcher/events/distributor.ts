import { FindingSeverity, FindingType } from '@fortanetwork/forta-bot'

import { CSFeeDistributor__factory } from '../../../generated/typechain'
import * as CSFeeDistributor from '../../../generated/typechain/CSFeeDistributor'
import { EventOfNotice } from '../../../shared/types'
import { ipfsLink } from '../../../utils/string'

const ICSFeeDistributor = CSFeeDistributor__factory.createInterface()

export function getCSFeeDistributorEvents(distributorAddress: string): EventOfNotice[] {
    return [
        {
            address: distributorAddress,
            abi: ICSFeeDistributor.getEvent('DistributionDataUpdated').format('full'),
            alertId: 'CSFEE-DISTRIBUTOR-DISTRIBUTION-DATA-UPDATED',
            name: 'ðŸ”µ CSFeeDistributor: Distribution data updated',
            description: (args: CSFeeDistributor.DistributionDataUpdatedEvent.OutputObject) =>
                `Total Claimable Shares: ${args.totalClaimableShares}\n` +
                `Tree Root: ${args.treeRoot}\n` +
                `Tree CID: ${ipfsLink(args.treeCid)}`,
            severity: FindingSeverity.Info,
            type: FindingType.Info,
        },
        {
            address: distributorAddress,
            abi: ICSFeeDistributor.getEvent('DistributionLogUpdated').format('full'),
            alertId: 'CSFEE-DISTRIBUTOR-DISTRIBUTION-LOG-UPDATED',
            name: 'ðŸ”µ CSFeeDistributor: Distribution log updated',
            description: (args: CSFeeDistributor.DistributionLogUpdatedEvent.OutputObject) =>
                `Log CID: ${ipfsLink(args.logCid)}`,
            severity: FindingSeverity.Info,
            type: FindingType.Info,
        },
    ]
}
