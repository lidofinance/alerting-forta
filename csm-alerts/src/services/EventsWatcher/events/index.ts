export * from './assetRecoverer'
export * from './distributor'
export * from './accounting'
export * from './pausable'
export * from './proxies'
export * from './module'
export * from './oracle'
export * from './roles'
