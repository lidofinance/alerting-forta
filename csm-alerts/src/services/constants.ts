import { DeployedAddresses } from '../shared/types'

export const DEPLOYED_ADDRESSES: DeployedAddresses = {
    CS_MODULE: '0xdA7dE2ECdDfccC6c3AF10108Db212ACBBf9EA83F',
    CS_ACCOUNTING: '0x4d72BFF1BeaC69925F8Bd12526a39BAAb069e5Da',
    CS_FEE_DISTRIBUTOR: '0xD99CC66fEC647E68294C6477B40fC7E0F6F618D0',
    CS_FEE_ORACLE: '0x4D4074628678Bd302921c20573EEa1ed38DdF7FB',
    CS_VERIFIER: '0x3Dfc50f22aCA652a0a6F28a0F892ab62074b5583',
    CS_EARLY_ADOPTION: '0x3D5148ad93e2ae5DedD1f7A8B3C19E7F67F90c0E',
    CS_GATE_SEAL: '0x16Dbd4B85a448bE564f1742d5c8cCdD2bB3185D0',
    LIDO_STETH: '0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84',
    BURNER: '0xD15a672319Cf0352560eE76d9e89eAB0889046D3',
    HASH_CONSENSUS: '0x71093efF8D8599b5fA340D665Ad60fA7C80688e4',
    STAKING_ROUTER: '0xFdDf38947aFB03C621C71b06C9C70bce73f12999',
    MULTICALL3: '0xcA11bde05977b3631167028862bE2a173976CA11',
    PROVER: '0xDCBDb89D30E921C73B496e9d20a61d283A5D2401',
}

export const ALIASES: Record<keyof DeployedAddresses, string> = {
    CS_MODULE: 'CSM',
    CS_ACCOUNTING: 'CSAccounting',
    CS_FEE_DISTRIBUTOR: 'CSFeeDistributor',
    CS_FEE_ORACLE: 'CSFeeOracle',
    CS_VERIFIER: 'CSAccounting',
    CS_EARLY_ADOPTION: 'CSEarlyAdoption',
    CS_GATE_SEAL: 'CSM Gate Seal',
    LIDO_STETH: 'Lido',
    BURNER: 'Burner',
    HASH_CONSENSUS: 'CSM HashConsensus',
    STAKING_ROUTER: 'StakingRouter',
    MULTICALL3: 'Multicall3',
    PROVER: 'Prover Wallet',
}

export const ORACLE_MEMBERS: { [key: string]: string } = {
    '0x140Bd8FbDc884f48dA7cb1c09bE8A2fAdfea776E': 'Chorus One',
    '0x404335BcE530400a5814375E7Ec1FB55fAff3eA2': 'Staking Facilities',
    '0x946D3b081ed19173dC83Cd974fC69e1e760B7d78': 'Stakefish',
    '0x007DE4a5F7bc37E2F26c0cb2E8A95006EE9B89b5': 'P2P',
    '0xc79F702202E3A6B0B6310B537E786B9ACAA19BAf': 'Chainlayer',
    '0x61c91ECd902EB56e314bB2D5c5C07785444Ea1c8': 'bloXroute',
    '0x1Ca0fEC59b86F549e1F1184d97cb47794C8Af58d': 'Instadapp', // Revoked one.
    '0x73181107c8D9ED4ce0bbeF7A0b4ccf3320C41d12': 'Instadapp',
    '0xe57B3792aDCc5da47EF4fF588883F0ee0c9835C9': 'MatrixedLink',
    '0xA7410857ABbf75043d61ea54e07D57A6EB6EF186': 'Kyber Network',
}
