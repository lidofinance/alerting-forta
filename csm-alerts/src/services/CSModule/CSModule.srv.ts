import { either as E } from 'fp-ts'
import { EventOfNotice, TransactionDto, handleEventsOfNotice, BlockDto } from '../../entity/events'
import { elapsedTime } from '../../utils/time'
import { Logger } from 'winston'
import { Finding } from '../../generated/proto/alert_pb'
import { ethers, filterLog, getEthersProvider } from 'forta-agent'
import CS_MODULE_ABI from '../../brief/abi/CSModule.json'
import BigNumber from 'bignumber.js'
import {
  BASIS_POINTS_DIVIDER,
  MAX_EMPTY_BATCHES_IN_THE_QUEUE,
  MAX_OPERATORS_WITH_SAME_MANAGER_OR_REWARD_ADDRESS,
  MAX_TARGET_SHARE_PERCENT_USED,
  MAX_VALIDATORS_IN_THE_QUEUE,
  ONE_DAY,
} from '../../utils/constants'
import { assertInvariant, getLogsByChunks } from '../../utils/utils'
import STAKING_ROUTER_ABI from '../../brief/abi/StakingRouter.json'
import { Config } from '../../utils/env/env'
import { NODE_OPERATOR_ADDED_EVENT_ABI } from '../../utils/events/cs_module_events'

export abstract class ICSModuleClient {
  public abstract getBlockByNumber(blockNumber: number): Promise<E.Either<Error, BlockDto>>
}

class Batch {
  value: BigNumber

  constructor(v: BigNumber) {
    this.value = v
  }

  noId(): bigint {
    return BigInt(this.value.toString()) >> BigInt(192)
  }

  keys(): bigint {
    return (BigInt(this.value.toString()) >> BigInt(128)) & BigInt('0xFFFFFFFFFFFFFFFF')
  }

  next(): bigint {
    return BigInt(this.value.toString()) & BigInt('0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF')
  }
}

interface StakingModule {
  stakingModuleAddress: string
  targetShare: number // in basis points
}

export class CSModuleSrv {
  private readonly name = 'CSModuleSrv'
  private readonly logger: Logger
  private readonly csModuleClient: ICSModuleClient
  private readonly csModuleAddress: string
  private readonly stakingRouterAddress: string
  private readonly csModuleEvents: EventOfNotice[]
  private lastDuplicateAddressesAlertTimestamp: number = 0
  private lastModuleShareIsCloseToTargetShareAlertTimestamp: number = 0
  private lastTooManyEmptyBatchesAlertTimestamp: number = 0
  private lastTooManyValidatorsAlertTimestamp: number = 0
  private addressCounts: { [address: string]: number } = {}
  private nodeOperatorAddedEvents: ethers.Event[] = []

  constructor(
    logger: Logger,
    csModuleClient: ICSModuleClient,
    csModuleAddress: string,
    stakingRouterAddress: string,
    csModuleEvents: EventOfNotice[],
  ) {
    this.logger = logger
    this.csModuleClient = csModuleClient
    this.csModuleAddress = csModuleAddress
    this.stakingRouterAddress = stakingRouterAddress
    this.csModuleEvents = csModuleEvents
  }

  public async initialize(currentBlock: number): Promise<Error | null> {
    const start = new Date().getTime()

    const config = new Config()

    const currBlock = await this.csModuleClient.getBlockByNumber(currentBlock)
    if (E.isLeft(currBlock)) {
      this.logger.error(elapsedTime(`Failed [${this.name}.initialize]`, start))
      return currBlock.left
    }

    const csModule = new ethers.Contract(this.csModuleAddress, CS_MODULE_ABI, getEthersProvider())
    const startBlock = config.csModuleInitBlock
    const csModuleNodeOperatorAddedFilter = csModule.filters.NodeOperatorAdded()
    this.nodeOperatorAddedEvents = await getLogsByChunks(
      csModule,
      csModuleNodeOperatorAddedFilter,
      startBlock,
      currentBlock,
    )

    this.logger.info(elapsedTime(`[${this.name}.initialize]`, start))
    return null
  }

  public getName(): string {
    return this.name
  }

  public handleEveryHundredOperatorsCreated(txEvent: TransactionDto) {
    const out: Finding[] = []

    const nodeOperatorAddedEvents = filterLog(txEvent.logs, NODE_OPERATOR_ADDED_EVENT_ABI, this.csModuleAddress)
    nodeOperatorAddedEvents.forEach((event) => {
      if (Number(event.args.nodeOperatorId) % 100 === 0 || Number(event.args.nodeOperatorId) === 69) {
        const f: Finding = new Finding()
        f.setName(`ðŸ”µ CSModule: Notable Node Operator creation`)
        f.setDescription(`Operator #${event.args.nodeOperatorId} was created.`)
        f.setAlertid('CS-MODULE-NOTABLE-NODE-OPERATOR-CREATION')
        f.setSeverity(Finding.Severity.INFO)
        f.setType(Finding.FindingType.INFORMATION)
        f.setProtocol('ethereum')

        out.push(f)
      }
    })

    return out
  }

  async handleTransaction(txEvent: TransactionDto): Promise<Finding[]> {
    const out: Finding[] = []

    const csModule = new ethers.Contract(this.csModuleAddress, CS_MODULE_ABI, getEthersProvider())
    // Important to get events of type ethers.Event specificly
    // therefore fetching current block
    const csModuleNodeOperatorAddedFilter = csModule.filters.NodeOperatorAdded()
    const [events] = await getLogsByChunks(
      csModule,
      csModuleNodeOperatorAddedFilter,
      txEvent.block.number, // start block,
      txEvent.block.number, // end block,
    )
    if (events) {
      this.nodeOperatorAddedEvents.push(events)
    }

    const csModuleFindings = handleEventsOfNotice(txEvent, this.csModuleEvents)
    const everyHundredOperatorsCreatedFindings = this.handleEveryHundredOperatorsCreated(txEvent)

    out.push(...csModuleFindings, ...everyHundredOperatorsCreatedFindings)

    return out
  }

  async getTotalActiveValidators(blockDto: BlockDto) {
    const stakingRouter = new ethers.Contract(this.stakingRouterAddress, STAKING_ROUTER_ABI, getEthersProvider())
    const [smDigests] = await stakingRouter.functions.getAllStakingModuleDigests({
      blockTag: blockDto.number,
    })
    let totalActiveValidators = 0
    for (const smDigest of smDigests) {
      const summary = smDigest.summary
      const totalDepositedValidators = (summary.totalDepositedValidators as BigNumber).toNumber()
      const totalExitedValidators = (summary.totalExitedValidators as BigNumber).toNumber()
      totalActiveValidators += totalDepositedValidators - totalExitedValidators
    }

    return totalActiveValidators
  }

  public async handleModuleShareIsCloseToTargetShare(
    blockDto: BlockDto,
    totalActiveValidators: number,
  ): Promise<Finding[]> {
    const out: Finding[] = []
    const now = blockDto.timestamp

    const csModule = new ethers.Contract(this.csModuleAddress, CS_MODULE_ABI, getEthersProvider())
    const stakingRouter = new ethers.Contract(this.stakingRouterAddress, STAKING_ROUTER_ABI, getEthersProvider())

    const summary = await csModule.functions.getStakingModuleSummary()
    const csTotalDepositedValidators = (summary.totalDepositedValidators as BigNumber).toNumber()
    const csTotalExitedValidators = (summary.totalExitedValidators as BigNumber).toNumber()
    const csTotalActiveValidators = csTotalDepositedValidators - csTotalExitedValidators

    const stakingModules = await stakingRouter.functions.getStakingModules()
    const csModuleInfo = stakingModules.find(
      (module: StakingModule) => module.stakingModuleAddress === this.csModuleAddress,
    )

    const currentShare = Math.ceil((csTotalActiveValidators / totalActiveValidators) * 100)
    const targetShare = Math.ceil(csModuleInfo.targetShare / BASIS_POINTS_DIVIDER)
    const percentUsed = Math.ceil((currentShare / targetShare) * 100)

    const timeSinceLastAlert = now - this.lastModuleShareIsCloseToTargetShareAlertTimestamp

    if (timeSinceLastAlert > ONE_DAY) {
      if (percentUsed >= MAX_TARGET_SHARE_PERCENT_USED) {
        const f: Finding = new Finding()
        f.setName(`ðŸŸ¢ CSModule: Module's share is close to the targetShare.`)
        f.setDescription(
          `The module's share is close to the target share (${percentUsed}% used). The module has ${csTotalActiveValidators} validators against ${totalActiveValidators} total. Target share: ${csModule.targetShare}`,
        )
        f.setAlertid('CS-MODULE-CLOSE-TO-TARGET-SHARE')
        f.setSeverity(Finding.Severity.LOW)
        f.setType(Finding.FindingType.INFORMATION)
        f.setProtocol('ethereum')

        out.push(f)

        this.lastModuleShareIsCloseToTargetShareAlertTimestamp = now
      }
    }
    return out
  }

  public async handleDuplicateAddresses(blockDto: BlockDto) {
    const out: Finding[] = []
    const now = blockDto.timestamp

    this.nodeOperatorAddedEvents.forEach((event) => {
      const addresses = [event.args?.managerAddress, event.args?.rewardAddress]
      addresses.forEach((address) => {
        if (address) {
          this.addressCounts[address] = (this.addressCounts[address] ?? 0) + 1
        }
      })
    })

    const timeSinceLastAlert = now - this.lastDuplicateAddressesAlertTimestamp

    if (timeSinceLastAlert > ONE_DAY) {
      Object.entries(this.addressCounts).forEach(([address, count]) => {
        if (count > MAX_OPERATORS_WITH_SAME_MANAGER_OR_REWARD_ADDRESS) {
          const f: Finding = new Finding()
          f.setName(
            `ðŸŸ¢ CSModule: More than ${MAX_OPERATORS_WITH_SAME_MANAGER_OR_REWARD_ADDRESS} operators have the same manager or reward address.`,
          )
          f.setDescription(`${count} operators have ${address} as manager or reward address.`)
          f.setAlertid('CS-MODULE-DUPLICATE-MANAGER-OR-REWARD-ADDRESS')
          f.setSeverity(Finding.Severity.LOW)
          f.setType(Finding.FindingType.INFORMATION)
          f.setProtocol('ethereum')

          out.push(f)

          this.lastDuplicateAddressesAlertTimestamp = now
        }
      })
    }
    return out
  }

  public async handleDepositQueueAlerts(blockDto: BlockDto) {
    const out: Finding[] = []
    const now = blockDto.timestamp

    const csModule = new ethers.Contract(this.csModuleAddress, CS_MODULE_ABI, getEthersProvider())
    const queueLookup: Map<bigint, bigint> = new Map()
    let emptyBatchCount = 0
    const queue = await csModule.depositQueue()
    let validatorsInQueue = 0
    let index = queue.head

    while (index < queue.tail) {
      const batchValue: BigNumber = await csModule.depositQueueItem(index)
      const batch = new Batch(batchValue)
      if (batch.next() === BigInt(0)) {
        break
      }
      const nodeOperatorId = batch.noId()
      const keysInBatch = batch.keys()
      const nodeOperator = await csModule.getNodeOperator(nodeOperatorId, { blockTag: blockDto.number })
      const depositableKeys = nodeOperator.depositableValidatorsCount

      const keysSeenForOperator = queueLookup.get(nodeOperatorId) || 0

      if (BigInt(keysSeenForOperator) >= depositableKeys) {
        emptyBatchCount++
      } else {
        queueLookup.set(nodeOperatorId, BigInt(keysSeenForOperator) + keysInBatch)
        validatorsInQueue += Number(keysInBatch)
      }

      index = batch.next()
    }

    const timeSinceLastEmptyBatchesAlert = now - this.lastTooManyEmptyBatchesAlertTimestamp

    if (timeSinceLastEmptyBatchesAlert > ONE_DAY) {
      if (emptyBatchCount > MAX_EMPTY_BATCHES_IN_THE_QUEUE) {
        const f: Finding = new Finding()
        f.setName(`ðŸŸ¢ CSModule: Too many empty batces in the deposit queue.`)
        f.setDescription(`${emptyBatchCount} empty batces in the deposit queue.`)
        f.setAlertid('CS-MODULE-TOO-MANY-EMPTY-BATCHES-IN-THE-QUEUE')
        f.setSeverity(Finding.Severity.LOW)
        f.setType(Finding.FindingType.INFORMATION)
        f.setProtocol('ethereum')

        out.push(f)

        this.lastTooManyEmptyBatchesAlertTimestamp = now
      }
    }

    const timeSinceLastMaxValidatorsAlert = now - this.lastTooManyValidatorsAlertTimestamp

    if (timeSinceLastMaxValidatorsAlert > ONE_DAY) {
      if (validatorsInQueue > MAX_VALIDATORS_IN_THE_QUEUE) {
        const f: Finding = new Finding()
        f.setName('ðŸŸ¢ CSModule: Too many validators in the queue.')
        f.setDescription(`CSModule has ${validatorsInQueue} keys in the deposit queue.`)
        f.setAlertid('CS-MODULE-TOO-MANY-VALIDATORS-IN-THE-QUEUE')
        f.setSeverity(Finding.Severity.LOW)
        f.setType(Finding.FindingType.INFORMATION)
        f.setProtocol('ethereum')

        out.push(f)

        this.lastTooManyValidatorsAlertTimestamp = now
      }
    }

    return out
  }

  async handleBlock(blockDto: BlockDto): Promise<Finding[]> {
    const start = new Date().getTime()
    const findings: Finding[] = []

    const csModule = new ethers.Contract(this.csModuleAddress, CS_MODULE_ABI, getEthersProvider())

    // Invariants
    let totalDepositedValidators = 0
    let totalExitedValidators = 0
    let depositableValidatorsCount = 0

    for (let noId = 0; noId < csModule.getNodeOperatorsCount(); noId++) {
      const noSummary = csModule.getNodeOperatorSummary(noId)

      if (!csModule.publicRelease) {
        assertInvariant(
          noSummary.totalAddedKeys > csModule.MAX_SIGNING_KEYS_PER_OPERATOR_BEFORE_PUBLIC_RELEASE,
          `Invariant failed: Node Operator ${noId} has more keys than allowed before public release.`,
          findings,
        )
      }

      const no = await csModule.getNodeOperator(noId)

      assertInvariant(
        !(no.totalAddedKeys >= no.totalDepositedKeys && no.totalDepositedKeys >= no.totalWithdrawnKeys),
        `Invariant failed: totalAddedKeys >= totalDepositedKeys >= totalWithdrawnKeys for Node Operator ${noId}`,
        findings,
      )
      assertInvariant(
        !(no.totalDepositedKeys <= no.totalVettedKeys && no.totalVettedKeys <= no.totalAddedKeys),
        `Invariant failed: totalDepositedKeys <= totalVettedKeys <= totalAddedKeys for Node Operator ${noId}`,
        findings,
      )
      assertInvariant(
        !(no.stuckValidatorsCount + no.totalWithdrawnKeys <= no.totalDepositedKeys),
        `Invariant failed: stuckValidatorsCount + totalWithdrawnKeys <= totalDepositedKeys for Node Operator ${noId}`,
        findings,
      )
      assertInvariant(
        !(no.depositableValidatorsCount + no.totalExitedKeys <= no.totalAddedKeys),
        `Invariant failed: depositableValidatorsCount + totalExitedKeys <= totalAddedKeys for Node Operator ${noId}`,
        findings,
      )
      assertInvariant(
        no.proposedManagerAddress === no.managerAddress,
        `Invariant failed: proposedManagerAddress should not equal managerAddress for Node Operator ${noId}`,
        findings,
      )
      assertInvariant(
        no.proposedRewardAddress === no.rewardAddress,
        `Invariant failed: proposedRewardAddress should not equal rewardAddress for Node Operator ${noId}`,
        findings,
      )
      assertInvariant(
        no.managerAddress === ethers.constants.AddressZero,
        `Invariant failed: managerAddress should not be zero address for Node Operator ${noId}`,
        findings,
      )
      assertInvariant(
        no.rewardAddress === ethers.constants.AddressZero,
        `Invariant failed: rewardAddress should not be zero address for Node Operator ${noId}`,
        findings,
      )

      const queue = await csModule.depositQueue()
      let index = queue.head
      const tail = queue.tail
      let enqueuedCount = BigInt(0)
      while (index < tail) {
        const batchValue: BigNumber = await csModule.depositQueueItem(index)
        const batch = new Batch(batchValue)
        if (batch.noId() === BigInt(noId)) {
          enqueuedCount += batch.keys()
        }
        index = batch.next()
        if (index === BigInt(0)) {
          break
        }
      }

      assertInvariant(
        no.enqueuedCount !== enqueuedCount,
        `Invariant failed: enqueuedCount mismatch for Node Operator ${noId}`,
        findings,
      )

      totalDepositedValidators += new BigNumber(noSummary.totalDepositedKeys.toString()).toNumber()
      totalExitedValidators += new BigNumber(noSummary.totalExitedKeys.toString()).toNumber()
      depositableValidatorsCount += new BigNumber(noSummary.depositableValidatorsCount.toString()).toNumber()
    }

    const smSummary = await csModule.functions.getStakingModuleSummary()
    assertInvariant(
      smSummary.totalDepositedValidators !== totalDepositedValidators ||
        smSummary.totalExitedValidators !== totalExitedValidators ||
        smSummary.depositableValidatorsCount !== depositableValidatorsCount,
      `Invariant failed: Global summary values do not match the accumulated node operator summaries.`,
      findings,
    )

    const moduleShareIsCloseToTargetShareFindings = this.handleModuleShareIsCloseToTargetShare(
      blockDto,
      await this.getTotalActiveValidators(blockDto),
    )
    const duplicateAddressesFindings = this.handleDuplicateAddresses(blockDto)
    const depositQueueFindings = this.handleDepositQueueAlerts(blockDto)

    findings.push(
      ...(await moduleShareIsCloseToTargetShareFindings),
      ...(await duplicateAddressesFindings),
      ...(await depositQueueFindings),
    )

    this.logger.info(elapsedTime(CSModuleSrv.name + '.' + this.handleBlock.name, start))

    return findings
  }
}
