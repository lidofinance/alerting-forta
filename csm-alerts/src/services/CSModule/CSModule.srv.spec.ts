import { DeploymentAddress, DeploymentAddresses } from '../../utils/constants.holesky'
import { expect } from '@jest/globals'
import { TransactionDto } from '../../entity/events'
import {
  CSModule__factory,
  CSAccounting__factory,
  CSFeeDistributor__factory,
  CSFeeOracle__factory,
} from '../../generated/typechain'
import { CSModuleSrv, ICSModuleClient } from './CSModule.srv'
import { getCSModuleEvents } from '../../utils/events/cs_module_events'
import * as Winston from 'winston'
import { ETHProvider } from '../../clients/eth_provider'
import { ethers } from 'forta-agent'
import { getFortaConfig } from 'forta-agent/dist/sdk/utils'
import promClient from 'prom-client'
import { Metrics } from '../../utils/metrics/metrics'

const TEST_TIMEOUT = 120_000 // ms

describe('CSModule event tests', () => {
  const chainId = 17000

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const address: DeploymentAddress = DeploymentAddresses

  const fortaEthersProvider = new ethers.providers.JsonRpcProvider(getFortaConfig().jsonRpcUrl, chainId)
  const csModuleRunner = CSModule__factory.connect(address.CS_MODULE_ADDRESS, fortaEthersProvider)
  const csAccountingRunner = CSAccounting__factory.connect(address.CS_ACCOUNTING_ADDRESS, fortaEthersProvider)
  const csFeeDistributorRunner = CSFeeDistributor__factory.connect(
    address.CS_FEE_DISTRIBUTOR_ADDRESS,
    fortaEthersProvider,
  )
  const csFeeOracleRunner = CSFeeOracle__factory.connect(address.CS_FEE_ORACLE_ADDRESS, fortaEthersProvider)

  const registry = new promClient.Registry()
  const m = new Metrics(registry, 'test_')

  const csModuleClient: ICSModuleClient = new ETHProvider(
    logger,
    m,
    fortaEthersProvider,
    csModuleRunner,
    csAccountingRunner,
    csFeeDistributorRunner,
    csFeeOracleRunner,
  )

  const csModuleSrv = new CSModuleSrv(
    logger,
    csModuleClient,
    address.CS_MODULE_ADDRESS,
    address.STAKING_ROUTER_ADDRESS,
    getCSModuleEvents(address.CS_MODULE_ADDRESS),
  )

  test(
    'ðŸŸ  CSModule: Target limit mode changed',
    async () => {
      const txHash = '0xd8bb4389a056be70fe20e3b6b903c3e7cdbf053610bd8d647c4e2fe49c94f8b6'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = await csModuleSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(1)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸ”´ CSModule: EL Rewards stealing penalty reported',
    async () => {
      const txHash = '0x45d08822a9e025d374ab182612a119d837a0869b0c343379025303f53c4c63be'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = await csModuleSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(1)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸ”µ CSModule: Notable Node Operator creation',
    async () => {
      const txHash = '0x87ece6668905293fd00c7eeff15ff685ed1d10810bc5cbba204f0881ab877be6'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = await csModuleSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(1)
    },
    TEST_TIMEOUT,
  )
})
