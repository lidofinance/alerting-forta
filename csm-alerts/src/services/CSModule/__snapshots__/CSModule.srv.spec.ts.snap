// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSModule event tests ðŸ”´ CSModule: EL Rewards stealing penalty reported 1`] = `
[
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "CS-MODULE-EL-REWARDS-STEALING-PENALTY-REPORTED",
      "ðŸ”´ CSModule: EL Rewards stealing penalty reported",
      "EL Rewards stealing penalty reported for Node Operator #37 with 0.004 ETH potentially stolen",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "37,0x9abf3544139d770248221e34fc90b40d9d2d526c7ae1638dc482bbcfae213c3b,3570000000000000",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;

exports[`CSModule event tests ðŸ”µ CSModule: Notable Node Operator creation 1`] = `
[
  {
    "array": [
      "ethereum",
      1,
      ,
      4,
      "CS-MODULE-NOTABLE-NODE-OPERATOR-CREATION",
      "ðŸ”µ CSModule: Notable Node Operator creation",
      "Operator #69 was created.",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": null,
  },
]
`;

exports[`CSModule event tests ðŸŸ  CSModule: Target limit mode changed 1`] = `
[
  {
    "array": [
      "ethereum",
      3,
      [],
      4,
      "CS-MODULE-TARGET-LIMIT-MODE-CHANGED",
      "ðŸŸ  CSModule: Target limit mode changed",
      "Target limit mode: 2 (forced mode)",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "3,2,5",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;
